Implementing Shor's algorithm using the Qrisp framework on Cat Qubits.ipynb

```diff
@@ -0,0 +1,770 @@
+    "    while(pow(g, e, N) != 1):\n",
+    "        e = e+1\n",
+    "    return e\n",
+    "\n",
+    "def classical_Shor(N):\n",
+    "    # returns the factors of N by using Shor's algorithm\n",
+    "    # with a classical brute-force period finding routine\n",
+    "    if N % 2 == 0:\n",
+    "        return (2, N // 2)\n",
+    "    if (is_prime(N)):\n",
+    "        return (1,N)\n",
+    "    \n",
+    "    while True:\n",
+    "        a = random.randint(2, N - 1)\n",
+    "        if math.gcd(a, N) != 1:\n",
+    "            return math.gcd(a, N), N // math.gcd(a, N)\n",
+    "        \n",
+    "        r = find_period_classical(a, N)\n",
+    "        \n",
+    "        if r % 2 == 0:\n",
+    "            x = pow(a, r // 2, N)\n",
+    "            p = math.gcd(x - 1, N)\n",
+    "            q = math.gcd(x + 1, N)\n",
+    "            if p != 1 and p != N:\n",
+    "                return (p, N // p)\n",
+    "            if q != 1 and q != N:\n",
+    "                return (q, N // q)"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 6,
+   "id": "e4a50a51",
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "Test passed!\n"
+     ]
+    }
+   ],
+   "source": [
+    "# Test the classical Shor's algorithm for N=10000\n",
+    "N=10000\n",
+    "check_list = list()\n",
+    "for i in range(2,N):\n",
+    "    (p,q) = classical_Shor(i)\n",
+    "    check = (p*q == i)\n",
+    "    if(check==False):\n",
+    "        print(i, (p,q))\n",
+    "else:\n",
+    "    print('Test passed!')"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 7,
+   "id": "01f21a4b",
+   "metadata": {},
+   "outputs": [
+    {
+     "data": {
+      "text/plain": [
+       "(3, 5)"
+      ]
+     },
+     "execution_count": 7,
+     "metadata": {},
+     "output_type": "execute_result"
+    }
+   ],
+   "source": [
+    "classical_Shor(15)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "a0d13a57",
+   "metadata": {},
+   "source": [
+    "The factorisation of 15 only requires small fractions of a second to be achieved with the brute-force algorithm.\n",
+    "\n",
+    "NOTE: you can use the `%%timeit` magic command to measure the time it takes to run a cell in a Jupyter notebook, although this will increase the actual time needed to execute the cell since multiple runs are performed. "
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "ce4d5d33",
+   "metadata": {},
+   "source": [
+    "#### Quantum Period finding\n"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 8,
+   "id": "9662c206-a627-4a79-b3e8-0abe336f9140",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "from sympy import continued_fraction_convergents, continued_fraction_iterator, Rational\n",
+    "\n",
+    "def get_r_candidates(approx):\n",
+    "    rationals = continued_fraction_convergents(continued_fraction_iterator(Rational(approx)))\n",
```

Comment:
nit-pick: Could yoy maybe write this over more lines to make it less long? I can show you how to use the black formatter

Reply:
Please do, I tried and I failed